### PR TITLE
Refactor the GPURenderPipelineDescriptor according to aspects

### LINKED ISSUE #936: Re-organize GPURenderPipelineDescriptor according to the structure of the GPU pipeline
Talked about this on the editors meeting. We'd want to reorganize the pipeline descriptors according to the following guidelines:
  - a descriptor has a field per stage (be it programmable or fixed-point stage)
  - each field is a "stage descriptor" (e.g. output targets go into the "output merger", index related things go first, primitive topology is between the vertex shader and rasterization, etc)
  - fields are ordered according to the real order, in which they are specified on the GPUs (input assembly -> vertex -> primitive assembly -> rasterization -> fragment -> output merger)

The benefit is - consistency in what's inside a sub-dictionary versus just a field on the top-level descriptor, and clear understanding on *when* things happen on the GPU.

### PR DESCRIPTION
**Pull request by kvark**
Groups the properties of a pipeline according to the aspects. Related things are kept together, even if they can affect multiple stages of a pipeline.

Closes #1307
Closes #936

I also want to re-order the sections of the pipeline, but this can be done as a follow-up in order to make this PR easier.

***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/1352.html" title="Last updated on Jan 21, 2021, 11:38 PM UTC (5a0b8d9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1352/a0d10e9...kvark:5a0b8d9.html" title="Last updated on Jan 21, 2021, 11:38 PM UTC (5a0b8d9)">Diff</a>

### CODE REVIEW

**Comment by toji** [comment]
On the last call you had an example that showed every possible descriptor property in one big struct. Could you link that again here or paste it into the comments? It's slightly tricky to get the complete picture from the various diffed chunks spread throughout the spec and it would serve as a useful reference for the intent.

**Comment by kvark** [comment]
```js
device.createRenderPipeline({
    vertex: {
        inputs: [{
            arrayStride: 16,
            stepMode: "vertex",
            attributes: [{
                offset: 0,
                format: "float2",
            }],
        }],
        module: module,
        entryPoint: "vs_main",
    },
    primitive: {
        topology: "triangle-strip",
        stripIndexFormat: "uint16",
        frontFace: "ccw",
        cullMode: "back",
    },
    depthStencil: {
        format: "depth24plus",
        depthWriteEnabled: true,
        depthCompare: "less",
        depthBias: 2,
        depthBiasSlopeScale: 1.0,
        depthBiasClamp: 8.0,
        stencilReadMask: 0xF,
        stencilWriteMask: 0x1,
        stencilFront: {
            compare: "greater",
            failOp: "keep",
            depthFailOp: "invert",
            passOp: "zero",
        },
        stencilBack: {},
    },
    multisample: {
        count: 4,
        mask: 0xF,
        alphaToCoverage: true,
    },
    fragment: {
        outputs: [{
            format: 'rgba8unorm',
            writeMask: 0xF,
            blend: {
                color: {
                    srcFactor: "one",
                    dstFactor: "src-alpha",
                    operation: "add",
                },
                alpha: {},
            },
        }],
        module: module,
        entryPoint: "fs_main",
    },
});
```

**Review by toji — COMMENTED** [review]
This is looking really good! Thanks!

File: spec/index.bs
Comment (kvark) [reply spec/index.bs:4307]: I understand the concern (input may seem like it refers to bind groups). And I think `inputBuffers`/`bufferInputs` may be equally confusing (buffers can also be in bind group). Not sure what a good name for this would be... we need to keep in mind that this is a "hot path", i.e. something that every single descriptor writer would need to type.

File: spec/index.bs
Comment (kvark) [reply spec/index.bs:3947]: We can do it for sure! I think we can also consider this to be a follow-up.

File: spec/index.bs
Comment (kvark) [reply spec/index.bs:4050]: Sure! Let's come back to this once we figure out what needs to be done with `inputs`, if anything

File: spec/index.bs
Comment (kvark) [reply spec/index.bs]: Please file an issue about this, it sounds interesting!
The `campDepth` is now moved to the end.

**Review by kvark — COMMENTED** [review]
Much detailed review, thank you!

File: spec/index.bs
Comment (toji) [reply spec/index.bs]: https://github.com/gpuweb/gpuweb/issues/1353

File: spec/index.bs
Comment (Kangz) [reply spec/index.bs:4310]: I independently had a similar concern when reading the full descriptor you posted above: `inputs` make it sound like it is a list of things that correspond to the `var<in>` in the shader, while it is a list of list of these things. That's why I have a slight preference for `buffers` because everyone understands what `vertex.buffers` means and its close relationship with `setVertexBuffers`.

File: spec/index.bs
Comment (Kangz) [inline spec/index.bs:3678]: nit: maybe rename to `compute` to match `fragment` and `vertex` in the `GPURenderPipelineDescriptor`?

**Review by Kangz — APPROVED** [review]
This change is amazing! Hopefully we don't need to move stuff around the `GPURenderPipelineDescriptor` a second time >_>

File: spec/index.bs
Comment (kvark) [reply spec/index.bs:3678]: oops, nice catch!

**Comment by kainino0x** [comment]
> This change is amazing! Hopefully we don't need to move stuff around the `GPURenderPipelineDescriptor` a second time >_>

FYI I want to consider a change after this one to change `alphaToCoverageEnabled: true` to `multisampleMode: "alpha-to-coverage"` because sample-rate shading would rightly go in that slot if we had it.

File: spec/index.bs
Comment (kainino0x) [reply spec/index.bs:4310]: I think it's okay to be slightly redundant in naming here for clarity. I don't think there will be SO many descriptors in a program that it would be arduous to type a few letters here. vertex.vertexBuffers, vertex.vertexData? I'm also fine with vertex.inputs as well, I don't have the same concern as Corentin though I have a tiny concern that it sounds like it would include uniform buffers.

**Review by kainino0x — APPROVED** [review]
looks great!

File: spec/index.bs
Comment (kvark) [reply spec/index.bs:4310]: ok, let's chat about the naming of these fields somewhere, and get it landed

**Comment by kvark** [comment]
In the last commit the changes are the following:
  1. tried to clean up the distinction between "xxxState" and "xxxDescriptor". Settled on the latter being used only for the top-level structures. Therefore, `VertexBufferLayoutDescriptor` is renamed to just `VertexBufferState`, and same for `VertexAttributeState`.
  1. renamed `GPUColorState` to `GPUColorTargetState`, since it's not a state of a "color"
  1. `inputs` became `inputBuffers` and `outputs` became `outputTargets`.

**Comment by kvark** [comment]
Also, rebased and squashed. It's very easy to run into conflicts with this.

**Comment by kainino0x** [comment]
> 1. Therefore, `VertexBufferLayoutDescriptor` is renamed to just `VertexBufferState`, and same for `VertexAttributeState`.

I like the word Layout, and think State isn't really important, could they just be VertexBufferLayout and VertexAttribute?

> Also, rebased and squashed. It's very easy to run into conflicts with this.

sg but a merge commit would have been easier for incremental review if possible.

**Comment by toji** [comment]
> inputs became inputBuffers and outputs became outputTargets

That's perfect! 😄 This change LGTM now.

> FYI I want to consider a change after this one to change alphaToCoverageEnabled: true to multisampleMode: "alpha-to-coverage"

Nit on Kai's follow up request: That would actually be `multisample.mode`, right?

**Review by toji — APPROVED** [review]

**Comment by kainino0x** [comment]
> Nit on Kai's follow up request: That would actually be `multisample.mode`, right?

Yeah, probably, unless we have a better name than `mode`.

**Comment by kvark** [comment]
Let's have the multisample discussion in a separate issue, perhaps? It sounds interesting, but I don't want it to drag this change.
@kainino0x your suggestion sounds appropriate, I've added the commit doing this.

As for the merge commit - there was a conflict already. I generally resolve conflicts by rebasing, but in this case it's much easier to resolve it on the whole thing than trying to resolve conflicts on each commit one by one. Just squashed to save time.

**Comment by Kangz** [comment]
> 3. inputs became inputBuffers and outputs became outputTargets.

I don't think the longer names were necessary, but unless more people feel the same I don't think my opinion should block this change.

**Review by kainino0x — APPROVED** [review]
LGTM modulo that final bikeshedding (I don't prefer the new names over inputs/outputs but they're okay)

> Let's have the multisample discussion in a separate issue, perhaps?

Yeah, was planning to open a PR after this lands.

**Comment by toji** [comment]
My sole reason for advocating for a name like `inputBuffers` is that it helps developers (especially ones new to the API) reason about what it describes better and disambiguates it from other types of shader input, but I think that's worthwhile.

> I don't think the longer names were necessary

We have some serious refactoring to do if "reduce the amount of typing required" is an API goal. 😉

File: spec/index.bs
Comment (toji) [inline spec/index.bs:4108]: Silly non-blocking nit: Given some of the other changes that are trying to reduce the usage of "Descriptor" this seems like it should be `GPUBlendState`

Comment (kvark) [reply spec/index.bs:4108]: right, right, I'll do it now!

**Comment by kvark** [comment]
@kainino0x oh, wait, I misunderstood. I thought you wanted these longer names, based on https://github.com/gpuweb/gpuweb/pull/1352#discussion_r561432534
If neither you or @Kangz are happy with these names, we should reconsider. What names would you want?

**Comment by kainino0x** [comment]
@kvark It's not that I have anything against slightly longer names, I just am not sure whether these are actually clearer - e.g. inputBuffers doesn't clarify that it doesn't include uniform buffers (though this probably doesn't matter that much). But if Brandon thinks so I think it's a good signal.

**Review by kainino0x — APPROVED** [review]

**Review by toji — APPROVED** [review]

**Comment by kainino0x** [comment]
> FYI I want to consider a change after this one to change `alphaToCoverageEnabled: true` to `multisampleMode: "alpha-to-coverage"` because sample-rate shading would rightly go in that slot if we had it.

Now that this PR landed I started looking into this. I learned that in fact alpha-to-coverage and sample-rate-shading are not exclusive with one another, so this idea doesn't make sense.

Sample-rate-shading means shading more than once per pixel, but can be fewer than once per sample (interpolation may occur for other samples, I think). In Vulkan I think the final coverage mask is the AND of all three mask values (pipeline sample mask, shader output sample mask, and alpha-to-coverage sample mask\*), then only the relevant bits of the final mask are used (those for the samples covered by the shader invocation).

\* though there is no guarantee about the alpha-to-coverage mask produced. According to #267, D3D12 disables alpha-to-coverage if SV_Coverage is used. I think this is why in WebGPU we don't allow alpha-to-coverage with SV_Coverage (added in #724), though given the leniency of the alpha-to-coverage mask I think we could simply allow it and stay within spec.

Additionally I learned that alpha-to-one is a separate (and optional) feature which is not turned on by enabling alpha-to-coverage.

**Comment by kainino0x** [comment]
> though given the leniency of the alpha-to-coverage mask I think we could simply allow it and stay within spec.

I should say, I don't think this is necessary. Being strict here is probably best as it sets expectations that the two aren't likely to actually work together.

**Pull request merged**
